### ISSUE TITLE
FP `explicit_auto_deref` on using never-to-any coercion with `&!`

### ISSUE THREAD

**Issue by kpreid** [rust-lang/rust-clippy#17713]
### Summary

`explicit_auto_deref` has a false positive when dereferencing an `&!`, which must be explicit since `!` doesn’t implement `Deref`. Perhaps the lint is firing on a coercion without checking that it is a deref coercion?

### Lint Name

explicit_auto_deref

### Reproducer

I tried this code:

```rust
#![warn(clippy::explicit_auto_deref)]
pub fn foo(input: &!) -> &u32 {
    *input
}
```

I saw this happen:

```
warning: deref which would be done by auto-deref
 --> src/lib.rs:3:5
  |
3 |     *input
  |     ^^^^^^ help: try: `input`
```

I expected to see this happen: No warning, because the code is already minimal.

### Version

```text
Clippy 0.1.100 (2026-09-07 cea272fa35)
```

### Additional Labels

@rustbot label I-suggestion-causes-error
